### PR TITLE
Add DecimalSelectiveReader

### PR DESCRIPTION
**Pull request by bhhari**
```
===NEW===
Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read              decimal(10,5)         true  avgt   20  0.234 ±  0.008   s/op
BenchmarkSelectiveStreamReaders.read              decimal(10,5)        false  avgt   20  0.471 ±  0.015   s/op
BenchmarkSelectiveStreamReaders.readAllNull       decimal(10,5)          N/A  avgt   20  0.002 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter    decimal(10,5)         true  avgt   20  0.277 ±  0.009   s/op
BenchmarkSelectiveStreamReaders.readWithFilter    decimal(10,5)        false  avgt   20  0.466 ±  0.018   s/op

Benchmark                                             Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readShortDecimalNoNull    avgt   60  0.481 ± 0.009   s/op
BenchmarkBatchStreamReaders.readShortDecimalWithNull  avgt   60  0.301 ± 0.005   s/op

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read             decimal(30,10)         true  avgt   20  0.288 ±  0.008   s/op
BenchmarkSelectiveStreamReaders.read             decimal(30,10)        false  avgt   20  0.446 ±  0.025   s/op
BenchmarkSelectiveStreamReaders.readAllNull      decimal(30,10)          N/A  avgt   20  0.002 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter   decimal(30,10)         true  avgt   20  0.350 ±  0.011   s/op
BenchmarkSelectiveStreamReaders.readWithFilter   decimal(30,10)        false  avgt   20  0.585 ±  0.036   s/op

Benchmark                                            Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readLongDecimalNoNull    avgt   60  0.645 ± 0.013   s/op
BenchmarkBatchStreamReaders.readLongDecimalWithNull  avgt   60  0.483 ± 0.010   s/op

===OLD===
Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read              decimal(10,5)         true  avgt   20  0.255 ±  0.008   s/op
BenchmarkSelectiveStreamReaders.read              decimal(10,5)        false  avgt   20  0.465 ±  0.022   s/op
BenchmarkSelectiveStreamReaders.readAllNull       decimal(10,5)          N/A  avgt   20  0.002 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter    decimal(10,5)         true  avgt   20  0.466 ±  0.013   s/op


# Run complete. Total time: 00:04:09

Benchmark                                             Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readShortDecimalNoNull    avgt   60  0.490 ± 0.014   s/op
BenchmarkBatchStreamReaders.readShortDecimalWithNull  avgt   60  0.344 ± 0.026   s/op



BenchmarkSelectiveStreamReaders.read             decimal(30,10)         true  avgt   20  0.385 ±  0.009   s/op
BenchmarkSelectiveStreamReaders.read             decimal(30,10)        false  avgt   20  0.611 ±  0.012   s/op
BenchmarkSelectiveStreamReaders.readAllNull      decimal(30,10)          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter   decimal(30,10)         true  avgt   20  0.430 ±  0.011   s/op
BenchmarkSelectiveStreamReaders.readWithFilter   decimal(30,10)        false  avgt   20  0.678 ±  0.008   s/op


Benchmark                                            Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readLongDecimalNoNull    avgt   60  0.636 ± 0.009   s/op
BenchmarkBatchStreamReaders.readLongDecimalWithNull  avgt   60  0.437 ± 0.013   s/op
```
====NO RELEASE NOTES====

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java [hunk 1]
```diff
@@ -38,6 +40,7 @@ private SelectiveStreamReaders() {}
     public static SelectiveStreamReader createStreamReader(
             StreamDescriptor streamDescriptor,
             Map<Subfield, TupleDomainFilter> filters,
+            List<OrcType> types,
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java]: I think a better change would to to store OrcType (not just OrcTypeKind) in the StreamDescriptor.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 2]
```diff
@@ -49,6 +49,9 @@
             "   CASE WHEN orderkey % 43  = 0 THEN null ELSE discount END as discount, \n" +
             "   CASE WHEN orderkey % 7 = 0 THEN null ELSE CAST(tax AS REAL) END AS tax_real, \n" +
             "   CASE WHEN linenumber % 2 = 0 THEN null ELSE (CAST(day(shipdate) AS TINYINT) , CAST(month(shipdate) AS TINYINT)) END AS ship_day_month, " +
+            "   CASE WHEN orderkey % 37 = 0 THEN null ELSE CAST(discount AS DECIMAL(20, 8)) END AS discount_long_decimal, " +
+            "   CASE WHEN orderkey % 41 = 0 THEN null ELSE CAST(tax AS DECIMAL(3, 2)) END AS tax_short_decimal, " +
+            "   CASE WHEN orderkey % 43 = 0 THEN null ELSE (CAST(discount AS DECIMAL(20, 8)), CAST(tax AS DECIMAL(20, 8))) END AS decimals, " +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: nit: decimals -> long_decimals

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java [hunk 3]
```diff
@@ -102,6 +112,7 @@ public static SelectiveStreamReader createStreamReader(
     public static SelectiveStreamReader createNestedStreamReader(
             StreamDescriptor streamDescriptor,
             int level,
+            List<OrcType> types,
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java]: this parameter is not used; let's remove it

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java [hunk 4]
```diff
@@ -169,7 +172,7 @@ else if (!filters.isEmpty()) {
         StreamDescriptor elementStreamDescriptor = streamDescriptor.getNestedStreams().get(0);
         Optional<Type> elementOutputType = outputType.map(type -> type.getTypeParameters().get(0));
 
-        this.elementStreamReader = createNestedStreamReader(elementStreamDescriptor, level + 1, Optional.ofNullable(this.listFilter), elementOutputType, hiveStorageTimeZone, systemMemoryContext);
+        this.elementStreamReader = createNestedStreamReader(elementStreamDescriptor, level + 1, types, Optional.ofNullable(this.listFilter), elementOutputType, hiveStorageTimeZone, systemMemoryContext);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: this change shouldn't be needed

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java [hunk 5]
```diff
@@ -82,7 +84,15 @@ public static SelectiveStreamReader createStreamReader(
                 return new StructSelectiveStreamReader(streamDescriptor, filters, requiredSubfields, outputType, hiveStorageTimeZone, systemMemoryContext);
             case MAP:
                 return new MapSelectiveStreamReader(streamDescriptor, filters, requiredSubfields, outputType, hiveStorageTimeZone, systemMemoryContext);
-            case DECIMAL:
+            case DECIMAL: {
+                DecimalType decimalType = DecimalType.createDecimalType(streamDescriptor.getOrcType().getPrecision().get(), streamDescriptor.getOrcType().getScale().get());
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java]: for consistency, I'd move the type creation into the reader:

```
            case DECIMAL: {
                if (streamDescriptor.getOrcType().getPrecision().get() <= MAX_SHORT_PRECISION) {
                    return new ShortDecimalSelectiveStreamReader(streamDescriptor, getOptionalOnlyFilter(type, filters), outputType, systemMemoryContext.newLocalMemoryContext(SelectiveStreamReaders.class.getSimpleName()));
                }
                else {
                    return new LongDecimalSelectiveStreamReader(streamDescriptor, getOptionalOnlyFilter(type, filters), outputType, systemMemoryContext.newLocalMemoryContext(SelectiveStreamReaders.class.getSimpleName()));
                }


    public AbstractDecimalSelectiveStreamReader(
            StreamDescriptor streamDescriptor,
            Optional<TupleDomainFilter> filter,
            Optional<Type> outputType,
            LocalMemoryContext systemMemoryContext)
    {
        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
        this.filter = filter.orElse(null);
        this.outputRequired = outputType.isPresent();
        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
        this.nonDeterministicFilter = this.filter != null && !this.filter.isDeterministic();
        this.nullsAllowed = this.filter == null || this.nonDeterministicFilter || this.filter.testNull();
        this.decimalType = DecimalType.createDecimalType(streamDescriptor.getOrcType().getPrecision().get(), streamDescriptor.getOrcType().getScale().get());

        this.nullBlock = this.decimalType.createBlockBuilder(null, 1).appendNull().build();
    }
```

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        }
+        else {
+            streamPosition = readWithFilter(positions, positionCount);
+        }
+
+        readOffset = offset + streamPosition;
+        return outputPositionCount;
+    }
+
+    protected int readAllNulls(int[] positions, int positionCount)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: make it private

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public void throwAnyError(int[] positions, int positionCount)
+    {
+    }
+
+    protected BlockLease newLease(Block block)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: make it private

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        return newLease(makeBlock(positionCount, includeNulls, nulls, values));
+    }
+
+    abstract void copyValues(int[] positions, int positionsCount, long[] valuesCopy, boolean[] nullsCopy);
+
+    abstract Block makeBlock(int positionCount, boolean includeNulls, boolean[] nulls, long[] values);
+
+    abstract long[] createValuesCopy(int positionCount);
+
+    protected void compactValues(int[] positions, int positionCount, boolean compactNulls)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: I think this method won't work for long decimals because it needs to know how many entries per position there are in `values` array.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+                    outputPositionCount++;
+                }
+            }
+            else {
+                Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+                Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+                int scale = (int) scaleStream.next();
+                dataStream.nextLongDecimal(decimal);
+                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
+                if (outputRequired) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: outputRequires is always true (since there is no filter); let's remove this `if`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    if (outputRequired) {
+                        nulls[outputPositionCount] = true;
+                    }
+                    outputPositionCount++;
+                }
+            }
+            else {
+                Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: - `unscaledDecimal` allocates new Slice and new byte buffer; these allocations happen for each row, but are not necessary; moving decimal and rescaledDecimal variables outside of the loop and reusing them in each iteration will avoid unnecessary allocations
- rescaledDecimal.getLong performs an unnecessary boundary check twice for each row (4 times for the case with filter); these checks are unnecessary and better avoided; 
- I made an *ugly* sketch of the above changes, re-run the bechmark and saw 1.4x improvement.

```
# Run complete. Total time: 00:02:17

Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read   decimal(30,10)         true  avgt   20  0.368 ± 0.027   s/op
BenchmarkSelectiveStreamReaders.read   decimal(30,10)        false  avgt   20  0.585 ± 0.010   s/op



# Run complete. Total time: 00:02:09

Benchmark                             (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read   decimal(30,10)         true  avgt   20  0.259 ± 0.006   s/op
BenchmarkSelectiveStreamReaders.read   decimal(30,10)        false  avgt   20  0.407 ± 0.008   s/op
```

Here is a sketch of the changes:

```
// no filter

        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        long address = rescaledDecimal.getAddress();

        // filter == null implies outputRequired == true
        int streamPosition = 0;
        for (int i = 0; i < positionCount; i++) {
            ...
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                byte[] data = (byte[]) rescaledDecimal.getBase();
                long low = unsafe.getLong(data, address);
                long high = unsafe.getLong(data, address + Long.BYTES);
                if (outputRequired) {
                    values[2 * outputPositionCount] = low;
                    values[2 * outputPositionCount + 1] = high;
                }
            ...
        }

// filter

        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
        long address = rescaledDecimal.getAddress();

        int streamPosition = 0;
        outputPositionCount = 0;
        for (int i = 0; i < positionCount; i++) {
            ...
            else {
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                byte[] data = (byte[]) rescaledDecimal.getBase();
                long low = unsafe.getLong(data, address);
                long high = unsafe.getLong(data, address + Long.BYTES);
                if (filter.testDecimal(low, high)) {
                    if (outputRequired) {
                        values[2 * outputPositionCount] = low;
                        values[2 * outputPositionCount + 1] = high;
                        if (nullsAllowed && presentStream != null) {
                            nulls[outputPositionCount] = false;
                        }
                    }
                    outputPositions[outputPositionCount] = position;
                    outputPositionCount++;
                }
            ...
        }
```

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java (hunk 8 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: will follow up on this

Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: agree, added code to handle 2 positions for Long and modified a test that will catch this.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
+            Block block = makeBlock(positionCount, nullsAllowed, nulls, values);
+            nulls = null;
+            values = null;
+            return block;
+        }
+
+        long[] valuesCopy = createValuesCopy(positionCount);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: consider replacing `createValuesCopy` and `ensureValuesCapacity` abstract methods with a member variable `valuesPerPosition`; let the subclasses provide the value (2 or 1) to the constructor.

`long[] valuesCopy = new long[valuesPerPosition * positionCount];`

otherwise, `createValuesCopy` needs a better name; currently it suggests that the method is making a copy, but it is only allocating memory

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount != outputPositionCount) {
+            compactValues(positions, positionCount, includeNulls);
+        }
+
+        return newLease(makeBlock(positionCount, includeNulls, nulls, values));
+    }
+
+    abstract void copyValues(int[] positions, int positionsCount, long[] valuesCopy, boolean[] nullsCopy);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: let's put all abstract methods next to each other

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    nulls[outputPositionCount] = true;
+                    outputPositionCount++;
+                }
+            }
+            else {
+                int scale = (int) scaleStream.next();
+                dataStream.nextLongDecimal(decimal);
+                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
+                long low = UnsafeSlice.getLongUnchecked(rescaledDecimal, 0);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: inline `low` and `high` variables; these are used only once

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+                long high = UnsafeSlice.getLongUnchecked(rescaledDecimal, Long.BYTES);
+                values[2 * outputPositionCount] = low;
+                values[2 * outputPositionCount + 1] = high;
+                if (nullsAllowed && presentStream != null) {
+                    nulls[outputPositionCount] = false;
+                }
+                outputPositionCount++;
+            }
+
+            outputPositions = positions;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: move this statement outside of the loop or remove it altogether

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: - since this is a `readNoFilter` method, filter is always `null`, nullsAllowed is always `true` and there is no need to compute outputPositionCount; a simplified version could look like this:

```
        int streamPosition = 0;
        for (int i = 0; i < positionCount; i++) {
            int position = positions[i];
            if (position > streamPosition) {
                skip(position - streamPosition);
                streamPosition = position;
            }

            if (presentStream != null && !presentStream.nextBit()) {
                nulls[i] = true;
            }
            else {
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                values[2 * i] = UnsafeSlice.getLongUnchecked(rescaledDecimal, 0);
                values[2 * i + 1] = UnsafeSlice.getLongUnchecked(rescaledDecimal, Long.BYTES);
                if (presentStream != null) {
                    nulls[i] = false;
                }
            }

            streamPosition++;
        }
        outputPositionCount = positionCount;
        return streamPosition;
```

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        super(streamDescriptor, filter, outputRequired, systemMemoryContext);
+    }
+
+    @Override
+    int readNoFilter(int[] positions, int positionCount)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: add `protected` modifier; ditto other methods

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+
+            if (positionIndex >= positionsCount) {
+                break;
+            }
+
+            nextPosition = positions[positionIndex];
+        }
+    }
+
+    void compactValues(int[] positions, int positionCount, boolean compactNulls)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: @Override; protected

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+            nextPosition = positions[positionIndex];
+        }
+
+        outputPositionCount = positionCount;
+    }
+
+    @Override
+    Block makeBlock(int positionCount, boolean includeNulls, boolean[] nulls, long[] values)
+    {
+        Block block = new Int128ArrayBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), values);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: nit: inline `block` variable

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java]: `filter` is always `null` and `nullsAllowed` is always `true`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+                    if (outputRequired) {
+                        nulls[outputPositionCount] = true;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java]: use `i` instead of `outputPositionCount`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    if (outputRequired) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java]: outputRequired is always `true`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+                if (outputRequired) {
+                    values[outputPositionCount] = rescale;
+                }
+                if (nullsAllowed && presentStream != null) {
+                    nulls[outputPositionCount] = false;
+                }
+                outputPositionCount++;
+            }
+
+            outputPositions = positions;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java]: move this outside of the loop or remove altogether

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractDecimalSelectiveStreamReader
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(AbstractDecimalSelectiveStreamReader.class).instanceSize();
+
+    protected static final Unsafe unsafe;
```
[the remark this answers is not in the document]
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: I feel this comment is on the outdated PR. I m using UnsafeSlice

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java [hunk 24]
```diff
@@ -28,6 +31,7 @@
 import java.util.Map;
 import java.util.Optional;
 
+import static com.facebook.presto.spi.type.Decimals.*;
```
[the remark this answers is not in the document]
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java]: same here outdated PR

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java (hunk 11 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: good idea

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java (hunk 14 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java]: yes its anyways init to positions

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(AbstractDecimalSelectiveStreamReader.class).instanceSize();
+
+    private final StreamDescriptor streamDescriptor;
+    protected final TupleDomainFilter filter;
+    protected final boolean nullsAllowed;
+    protected final boolean outputRequired;
+    protected final LocalMemoryContext systemMemoryContext;
+    protected final boolean nonDeterministicFilter;
+    protected final Block nullBlock;
+    protected final Integer scale;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: Integer -> int

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+    private final int valuesPerPosition;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: it: put all private variables next to each other

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java [hunk 27]
```diff
@@ -0,0 +1,337 @@
...
+    private boolean valuesInUse;
+
+    public AbstractDecimalSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            Optional<Type> outputType,
+            LocalMemoryContext systemMemoryContext,
+            int valuesPerPosition)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java]: "stream is null" -> "streamDescriptor is null"

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                nulls[i] = true;
+            }
+            else {
+                int scale = (int) scaleStream.next();
+                long rescale = Decimals.rescale(dataStream.nextLong(), scale, this.scale);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ShortDecimalSelectiveStreamReader.java]: nit: inline `rescale` variable

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 29]
```diff
@@ -309,7 +310,7 @@ public void testRoundTrip(Type type, List<?> readValues)
         testRoundTrip(type, readValues, ImmutableList.of());
     }
 
-    public void testRoundTrip(Type type, List<?> readValues, TupleDomainFilter...filters)
+    public void testRoundTrip(Type type, List<?> readValues, TupleDomainFilter... filters)
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: unnecessary change

Comment (bhhari) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: that was auto format when I did cmd+alt+L

**Comment by bhhari** [comment]
@mbasmanova  updated the commit with the latest benchmark data
